tensorflow_addons/metrics/kendalls_tau.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements Kendall's Tau metric and loss."""
+
+import warnings
+
+import numpy as np
+import tensorflow as tf
+from tensorflow_addons.metrics.utils import MeanMetricWrapper
+from tensorflow_addons.utils.types import TensorLike
+
+
+def _iterative_mergesort(y: TensorLike, aperm: TensorLike) -> (tf.int32, tf.Tensor):
+    """Non-recusive mergesort that counts exchanges.
+
+    Args:
+        y: values to be sorted.
+        aperm: original ordering.
+
+    Returns:
+        A tuple consisting of a int32 scalar that counts the number of
+        exchanges required to produce a sorted permutation, and a tf.int32
+        Tensor that contains the ordering of y values that are sorted.
+    """
+    exchanges = 0
+    num = tf.size(y)
+    k = tf.constant(1, tf.int32)
+    while tf.less(k, num):
```

Comment:
Since we are a general purpose library, we need to write everything so that it does not presume eager execution.  Minimally this means no python controlflow (eg, `for`, `while`, and `if` are not allowed).

More generally, this has a code nature that would be quite inefficient in a SIMD regime.  Perhaps we could examine the full quadratic comparison? Eg,

`0.5 * tf.math.reduce_mean(score[..., tf.newaxis] < score[..., tf.newaxis, :], axis=-1)`

In general this would be quite large. I therefore recommend extending this idea to "chunks" using a `tf.while_loop` to reduce them. In so doing we can handle many cases in memory yet also bound the memory by way of chunks.

Thoughts?

Reply:
Only after I finished writing this did I fully examine the implementation of the AUC metrics in Keras, which use a bucketized approximation, not that dissimilar to the two-dimensional bucket approach presented in https://arxiv.org/abs/1712.01521 which would be much more amenable to streaming. (OTOH, I'm unclear why _Metrics_ need to be streaming, usually the classifier labels is a space much smaller than the input features.)

I also note that I should have looked at the completely rewritten scipy implementation which works differently, but still requires sorting by both of the full inputs which seems largely incompatible with tf Metrics design.
https://github.com/scipy/scipy/blob/01d8bfb6f239df4ce70c799b9b485b53733c9911/scipy/stats/stats.py#L4452

I suspect a reasonable thing to do here is to probably fork this into two separate efforts, one for tfp which would focus on the explicit exact solution with potential backoff to approximate, and a tensorflow addons submission that focuses on implementing a streaming approximation. I think I can manage that, but it's going to take some time - and at present I don't have a good sense of which of these would be easier to do first.

Reply:
Hi - sorry I'm back to this finally - I think from the discussion this PR should be moved to tfp,
most likely in the https://github.com/tensorflow/probability/tree/master/tensorflow_probability/python/stats directory.

However, I think I need to make a significant redesign to comply with the suggestion of removing the control flow.

I'm not quite able to distill what the consensus is here for next steps, maybe I should do all of these:

- Make a tfp PR?
- Reimplement this as a streaming algorithm per the previously mentioned arxiv paper.
- Change this current PR to use tf.while(), possibly with chunks.